patch-upsert-v5.js

```diff
@@ -0,0 +1,568 @@
+          }
+        }
+      }
+    }
+
+    let onDuplicateKeyUpdate = '';
+
+    if (this._dialect.supports.inserts.updateOnDuplicate && options.updateOnDuplicate) {
+      if (this._dialect.supports.inserts.updateOnDuplicate == ' ON CONFLICT DO UPDATE SET') { // postgres / sqlite
+        // If no conflict target columns were specified, use the primary key names from options.upsertKeys
+        const conflictKeys = options.upsertKeys.map(attr => this.quoteIdentifier(attr));
+        const updateKeys = options.updateOnDuplicate.map(attr => `${this.quoteIdentifier(attr)}=EXCLUDED.${this.quoteIdentifier(attr)}`);
+        onDuplicateKeyUpdate = ` ON CONFLICT (${conflictKeys.join(',')}) DO UPDATE SET ${updateKeys.join(',')}`;
+      } else {
+        const valueKeys = options.updateOnDuplicate.map(attr => `${this.quoteIdentifier(attr)}=VALUES(${this.quoteIdentifier(attr)})`);
+        onDuplicateKeyUpdate += `${this._dialect.supports.inserts.updateOnDuplicate} ${valueKeys.join(',')}`;
+      }
+    }
+
+    const replacements = {
+      ignoreDuplicates: options.ignoreDuplicates ? this._dialect.supports.inserts.ignoreDuplicates : '',
+      onConflictDoNothing: options.ignoreDuplicates ? this._dialect.supports.inserts.onConflictDoNothing : '',
+      attributes: fields.join(','),
+      output: outputFragment,
+      values: values.join(','),
+      tmpTable
+    };
+
+    valueQuery = `${tmpTable}INSERT${replacements.ignoreDuplicates} INTO ${quotedTable} (${replacements.attributes})${replacements.output} VALUES (${replacements.values})${onDuplicateKeyUpdate}${replacements.onConflictDoNothing}${valueQuery}`;
+    emptyQuery = `${tmpTable}INSERT${replacements.ignoreDuplicates} INTO ${quotedTable}${replacements.output}${onDuplicateKeyUpdate}${replacements.onConflictDoNothing}${emptyQuery}`;
+
+    // Mostly for internal use, so we expect the user to know what he's doing!
+    // pg_temp functions are private per connection, so we never risk this function interfering with another one.
+    if (this._dialect.supports.EXCEPTION && options.exception) {
+      const dropFunction = 'DROP FUNCTION IF EXISTS pg_temp.testfunc()';
+
+      if (returningModelAttributes.length === 0) {
+        returningModelAttributes.push('*');
+      }
+
+      const delimiter = `$func_${uuidv4().replace(/-/g, '')}$`;
+      const selectQuery = `SELECT (testfunc.response).${returningModelAttributes.join(', (testfunc.response).')}, testfunc.sequelize_caught_exception FROM pg_temp.testfunc();`;
+
+      options.exception = 'WHEN unique_violation THEN GET STACKED DIAGNOSTICS sequelize_caught_exception = PG_EXCEPTION_DETAIL;';
+      valueQuery = `CREATE OR REPLACE FUNCTION pg_temp.testfunc(OUT response ${quotedTable}, OUT sequelize_caught_exception text) RETURNS RECORD AS ${delimiter} BEGIN ${valueQuery} RETURNING * INTO response; EXCEPTION ${options.exception} END ${delimiter} LANGUAGE plpgsql; ${selectQuery} ${dropFunction}`;
```

Comment:
cockroachdb doesn't support user-defined functions, so i wonder if we need this? i see that this won't be executed anyway since `supports.EXCEPTION` is set to false.

Reply:
You are right! I copied parts of the code from Sequelize v6 and made some changes until I got this part working, but surely the code could be simplified/improved.

I suggest we create another issue to track a (near-future) refactor of this part of the code, but I'd like to start working on the CI first, to see what else is working/not-working, which could bring more insight on further refactor that could be needed.

Reply:
Thanks! I agree -- simplifying this part of the code doesn't need to be a priority. Especially since this workaround is not needed for v6, so we won't need to maintain it after we stop v5 support.